### PR TITLE
ARROW-15154: [R] Expose ReferencedBufferSize to R

### PR DESCRIPTION
**Pull request by vibhatha**
In this PR, the `ReferencedBufferSize` functionality exposed in R API 

This is integrated for 

- [x] Array
- [x] ChunkedArray
- [x] RecordBatch
- [x] Table

### CODE REVIEW

File: r/src/array.cpp [hunk 1]
```diff
@@ -288,4 +289,10 @@ bool Array__Same(const std::shared_ptr<arrow::Array>& x,
                  const std::shared_ptr<arrow::Array>& y) {
   return x.get() == y.get();
 }
+
+// [[arrow::export]]
+int64_t Array__ReferencedBufferSize(const std::shared_ptr<arrow::Array>& x) {
+  auto array = arrow::internal::checked_cast<const arrow::Array*>(x.get());
+  return ValueOrStop(arrow::util::ReferencedBufferSize(*array));
```
Comment (pitrou) [inline r/src/array.cpp:296]: The checked_cast is pointless here, the type is the same left and right.
```suggestion
  return ValueOrStop(arrow::util::ReferencedBufferSize(*x));
```

File: r/src/chunkedarray.cpp [hunk 2]
```diff
@@ -143,4 +144,10 @@ std::shared_ptr<arrow::ChunkedArray> ChunkedArray__from_list(cpp11::list chunks,
   return std::make_shared<arrow::ChunkedArray>(std::move(vec));
 }
 
+// [[arrow::export]]
+int64_t ChunkedArray__ReferencedBufferSize(const std::shared_ptr<arrow::ChunkedArray>& x) {
+  auto chunked_array = arrow::internal::checked_cast<const arrow::ChunkedArray*>(x.get());
```
Comment (pitrou) [inline r/src/chunkedarray.cpp:149]: Same here.

File: r/src/recordbatch.cpp [hunk 3]
```diff
@@ -25,6 +25,8 @@
 #include <arrow/ipc/writer.h>
 #include <arrow/type.h>
 #include <arrow/util/key_value_metadata.h>
+#include <arrow/util/byte_size.h>
```
Comment (pitrou) [inline r/src/recordbatch.cpp:28]: Can you keep these includes alphabetically-ordered?

File: r/src/recordbatch.cpp [hunk 4]
```diff
@@ -305,4 +307,10 @@ std::shared_ptr<arrow::RecordBatch> RecordBatch__from_arrays(SEXP schema_sxp, SE
   return arrow::RecordBatch::Make(schema, num_rows, arrays);
 }
 
+// [[arrow::export]]
+int64_t RecordBatch__ReferencedBufferSize(const std::shared_ptr<arrow::RecordBatch>& x) {
+  auto batch = arrow::internal::checked_cast<const arrow::RecordBatch*>(x.get());
```
Comment (pitrou) [inline r/src/recordbatch.cpp:312]: Here too.

File: r/src/table.cpp [hunk 5]
```diff
@@ -21,6 +21,7 @@
 #include <arrow/array/array_base.h>
 #include <arrow/table.h>
 #include <arrow/util/key_value_metadata.h>
+#include <arrow/util/byte_size.h>
```
Comment (pitrou) [inline r/src/table.cpp:24]: Alphebetical ordering here too.

File: r/src/table.cpp [hunk 6]
```diff
@@ -283,4 +284,10 @@ std::shared_ptr<arrow::Table> Table__from_record_batches(
   return tab;
 }
 
+// [[arrow::export]]
+int64_t Table__ReferencedBufferSize(const std::shared_ptr<arrow::Table>& x) {
+  auto table = arrow::internal::checked_cast<const arrow::Table*>(x.get());
```
Comment (pitrou) [inline r/src/table.cpp:289]: Casting here too.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: r/src/recordbatch.cpp (hunk 3 above)
Comment (vibhatha) [reply r/src/recordbatch.cpp:28]: Sure, will update this.

**Comment by vibhatha** [comment]
@pitrou thank you for the review comments. I will address them :)

**Review by pitrou — APPROVED** [review]

**Review by thisisnic — COMMENTED** [review]
Looks great overall - thanks for this PR @vibhatha!  Just a couple of comments before I approve it:

1. The linter is complaining about some minor formatting things; you can fix these by running the styler on the files you have changed.  Instructions are in here: https://arrow.apache.org/docs/dev/r/articles/developers/workflow.html

2. Point 1 above may be redundant though as I don't think the unit tests you have added are needed - you're just exposing C++ functionality rather than adding to it or changing it in R, so I'd actually probably remove those tests entirely.  Looking at other similar methods (e.g. `type_id`) you'll see we don't test these in R.

**Comment by vibhatha** [comment]
@thisisnic thank you very much for the review. I will update the PR.

**Comment by vibhatha** [comment]
@thisisnic I updated the PR

**Review by thisisnic — APPROVED** [review]
Fantastic, thanks for this PR! :)

**Comment by vibhatha** [comment]
Thanks a lot @thisisnic ☺️

**Pull request closed without merging**